src/SubHask/Algebra/Vector.hs

```diff
@@ -1538,104 +1531,181 @@ type instance SetElem (ACCVector (bknd::Backend) n r) b = ACCVector (bknd::Backe
+-- where
+--
+--     {-# INLINE dim #-}
+--     dim (ACCVector v) = A.length v
 
-data ACCMatrix' b (m::k1) (n::k2) r where
-  Id    :: {-#UNPACK#-}!r              -> ACCMatrix' b m m r
-  Diag  :: {-#UNPACK#-}!(ACCVector b m r)  -> ACCMatrix' b m m r
-  Mat   :: {-#UNPACK#-}!(ACCMatrix b m n r)  -> ACCMatrix' b m n r
 
 
-type instance Scalar (ACCMatrix' b r m n) = Scalar r
-type instance (ACCMatrix' b r m n)><r = ACCMatrix b r m n
+instance
+    ( P.Num (A.Exp r)
+    , Monoid r
+    , ValidLogic r
+    , ValidACCVector b n r
+    , A.Elt r
+    , IsScalar r
+    , FreeModule r
+    ) => IxContainer (ACCVector b (n::Symbol) r)
+        where
+    --(!) also returns A.Exp r where r is expected ...
+    -- {-# INLINE (!) #-}
+    -- (!) (ACCVector v) i =  v A.! (A.index1 (A.lift i))
+    -- {-# INLINABLE  toIxList #-}
+    -- toIxList v = P.zip [0..] $ go (dim v-1) []
+    --     where
+    --         go (-1) xs = xs
+    --         go    i xs = go (i-1) (v!i : xs)
+    -- fails on f with:
+    -- Couldn't match type ‘Int’ with ‘A.Exp A.DIM1’
+    -- Expected type: A.Exp A.DIM1 -> A.Exp r -> A.Exp b1
+    --   Actual type: Index (ACCVector b n r)
+    --                -> Elem (ACCVector b n r) -> b1
+    -- {-# INLINABLE imap #-}
+    -- imap f (ACCVector v) = ACCVector (A.zipWith f (A.generate (A.shape v) P.id) v)
+
+    type ValidElem (ACCVector b n r) e = (ClassicalLogic e, IsScalar e, FiniteModule e, ValidACCVector b n e)
+
+
+instance (Eq r, Monoid r, ValidACCVector b n r) => Eq_ (ACCVector b (n::Symbol) r) where
+    -- returns A.Exp Bool . . . .
+    -- {-# INLINE (==) #-}
+    -- (ACCVector v2) == (ACCVector v1) = (A.lift v1) A.==* (A.lift v2)
+
+--Will acc arrays support Logic r~Bool?
+instance
+    ( ValidACCVector b n r
+    , P.Num (A.Exp r)
+    , ExpField r
+    , Normed r
+    , Ord_ r
+    , Logic r~Bool
+    , IsScalar r
+    , VectorSpace r
+    ) => Metric (ACCVector b (n::Symbol) r)
+        -- where
+
+    -- also seems to wan to deduce r ~ A.Acc (Scalar r)
+    -- {-# INLINE[2] distance #-}
+    -- distance (ACCVector v1) (ACCVector v2) = {-# SCC distance_ACCVector #-}let
+    --   dmag = A.zipWith (P.-) v1 v2
+    --   dsq = A.zipWith (P.*) dmag dmag
+    --   drt = A.sqrt (A.sum dsq)
+    --   in drt
+
+    --what is distanceUB tring to do?
+    -- {-# INLINE[2] distanceUB #-}
+    -- distanceUB v1 v2 ub = {-# SCC distanceUB_ACCVectorr #-}let
+    --   ub2 = ub*ub
+    --   dmag = A.zipWith (P.-) v1 v2
+    --   dsq = A.zipWith (P.*) dsq dsq
+    --   drt = A.map A.sqrt dsq
+    --   in expression
+
+instance (VectorSpace r, ValidACCVector b n r, IsScalar r, ExpField r) => Normed (ACCVector b (n::Symbol) r) where
+    -- {-# INLINE size #-}
+    -- size (ACCVector v1) = let
+    --   sq = A.zipWith (P.*) v1 v1
+    --   s = A.fold (P.+) 0.0 sq
+    --   in A.sqrt s
+
+
+-- Could not deduce (P.Floating (A.Exp r))
+--   arising from the superclasses of an instance declaration
```

Comment:
Since all of the instances are going to required these prelude constraints, I would add them to the `ValidACCVector` type you made.